### ISSUE TITLE
Feature Request:  Report additional image values 

### ISSUE THREAD

**Issue by profxj** [ejeschke/ginga#434]
I would like to add a plug-in (or use some other approach, e.g. WCS) that would
allow the user to:

- Load an additional ndarray (maybe more than one) with the same size as the display image
- Have Ginga report the values in this additional image as the cursor changes in x,y.  

My specific application is for 2D spectroscopy where I have a flux image and a
wavelength image of the same size.  I wish to display the flux in a Ginga channel
and have its flux Values reported along with the wavelength at the cursor x,y.

One thought is to overload the WCS as I am not using that functionality (i.e.
I have no WCS) or to simply use WCS if I can create an object that does what
I want (I'll bet I can).   

Advice before I dig in would be most appreciated.

**Comment by ejeschke** [comment]
@profxj, you are right that either a plugin or a custom WCS (or both) would be a good approach. I would suggest taking a look at the [`Cursor` plugin](https://github.com/ejeschke/ginga/blob/master/ginga/rv/plugins/Cursor.py)--an easy way would be to copy that (or subclass!) and replace the reporting done there with your own. Custom WCS is also fairly straightforward.

**Comment by profxj** [comment]
Ok, I think I have found an implementation that works well and only needs a new WCS class.
You can see the basic idea here:

https://github.com/ejeschke/ginga/compare/master...profxj:xcursor

@ejeschke , if you like the approach I'll polish it into a proper PR.

**Comment by pllim** [comment]
Is this "pseudo WCS" widely used or very specific to your use case? If the latter, it might be less confusing to live in your own package rather than here.

**Comment by profxj** [comment]
I feel that it could become widely used, and I'd add docs to help advertise its 
functionality.

**Comment by ejeschke** [comment]
@profxj, sorry for my late reply, I'm a bit swamped at work.

I think my main concern is adding the custom WCS directly to `wcsmod`.  It seems like it would be straightforward to define the custom wcs in your plugin (by simply importing `BaseWCS` from `wcsmod`).  If this wcs becomes widely and generically used we could consider to move it later.

It's also not a bad idea to consider distributing the new plugin separately using the new [ginga plugin template](https://github.com/ejeschke/ginga-plugin-template).  With the current code, if they install the plugin it will automatically show up in the menus.

But if you think this will be able to be used by a  fair number of current instruments/telescopes I am willing to add it to the stock distribution.  But "we" (collectively, ginga developers) probably need to start thinking about how to organize the proliferating set of plugins (see #310).  Probably we need to add a separate configuration step to enable/disable plugins similar to what Glue does. 

Were you thinking to replace the current `Cursor` with something that does the current behavior as well as handling the new kind of files or only activating the plugin when certainly kinds of files are opened, or my manually activating the plugin?

**Comment by pllim** [comment]
I am still not convinced that "pseudo WCS" will be widely used. So, I'll need more convincing.

**Comment by profxj** [comment]
No problem @ejeschke, @pllim  .  Here are a few additional comments to yours:

The proposed change does not involve a new (or edited) plugin.   It is only
a new subclass in wcsmod.  And I'd be happy to put it elsewhere but this
seems the sensible spot.

As for widely used,  I have now wrapped Ginga into our PYPIT data reduction 
pipeline:  http://pypit.readthedocs.io/en/latest/
In principle, anyone that winds up using this pipeline (it will be at least 10 people,
it could become 100+) may wind up using Ginga to view various pipeline outputs.  

If you should prefer the ImgWCS subclass not be in the main Ginga distribution, that is fine.  
I could just put it in PYPIT.   However, I probably also need to be able 
to modify the pseudo WCS image via RC which would mean sub-classing the existing RC
plugin in PYPIT, etc.  It just seems a bit unfortunate to fork off at this stage...

**Comment by ejeschke** [comment]
@profxj , that is great to hear about your use of Ginga in PYPIT.  Please don't take offense at questioning about "widely used".  Here at Subaru, where Ginga was originally written, we maintain our own private set of plugins and utility modules for use with the instances of Ginga we run on our telescope--because we expect that they are mostly applicable to our instrument files and workflow and not particularly relevant to others.  I believe @pllim does the same for some plugins with their use at STScI.  When we modify the "reference viewer" plugins it is always with an eye to fairly generic use.

But it is a primary goal with Ginga to make it easy to use the stock reference viewer with your own customizations.  I think it should be no problem to make it so that PYPIT can invoke Ginga and have it load your custom WCS.  I make some suggestions in the next comment.

**Comment by ejeschke** [comment]
One thing you can do is set the `GINGA_HOME` environment variable, which will tell the reference viewer to look in that directory for configuration files.  This is useful because then you can pre-seed it with configuration files customized for your pipeline.  One of the configuration files that you can put in that directory is `ginga_config.py`.  If this file is present Ginga will try to run three functions in it: `init_config(setup)`, `pre_gui_config(fv)` and `post_gui_config(fv)`. 

`init_config(setup)` can be used to manipulate the "ginga boot setup class"--basically it runs before the GUI toolkit has been detected and can be used to add paths, adjust preferences, add or remove plugins to the initialization lists, etc.

`pre_gui_config(fv)` runs before the GUI has been configured, but after the Ginga "shell" has been created (the shell is the thing that gets passed around to all the plugins as "fv").

And fiinally, `post_gui_config(fv)` is called after the plugins have been loaded, the GUI has been created, initial channels made, but just prior to turning control over to the user.

You could use any of these functions, particularly the `init_config` or `pre_gui_config` to load your WCS.

**Comment by profxj** [comment]
This is helpful.   It would be a bit more effort for our users to maintain
GINGA_HOME and a file within it but not much. 

And I agree with your approach to minimize modifications to the
stock reference viewer.   That said, the only Plugin I am intending
to touch (sometime soon) is the RC (which I have previously).  I can go 
forth and create an PYPIT_RC or x_RC plugin but it will be >95%
a duplicate of RC (I guess I'd subclass it).

Really the key question I am posing for now is whether you would be ok with 
a new wcspkg in Ginga which, like BareBones, is non-standard and whether it should
go in the wcsmod.py module.

**Comment by ejeschke** [comment]
@profxj , I think we should keep it out of `wcsmod` for now.  If it does become widely applicable we can consider putting it in there.  And we can (should) make any modifications to `wcsmod` necessary to register it from your own code bits.

As to `BareBones`, I've been thinking about deprecating/removing that for a while now.  It is only really there to handle the case where a user didn't have a better WCS package installed and it only really handles TAN projection.  Probably the thing to do is just report pixel coordinates and complain that the user needs to install a WCS package...

**Comment by ejeschke** [comment]
> It would be a bit more effort for our users to maintain `GINGA_HOME` and a file within it but not much.

Can't the startup software set this variable, or does the package not work like that?  You could also just have an initialization command that writes a set of files into the user's `$HOME/.ginga` (which is the default if `GINGA_HOME` is not set.

**Comment by ejeschke** [comment]
BTW, I think once you take the plunge to maintaining your own plugin(s) there is a lot of freedom there to make more interesting and extensive mods.  And subclassing is a good way to go if you are just adding a bit of functionality.

**Comment by ejeschke** [comment]
@profxj, is there a specific combination of header keywords that would proscribe the use of the pseudo WCS?

**Comment by profxj** [comment]
Yes, I'm using WCS-xIMG for now.
